Add Counterstrike mission Fall of Greece 1: Personal War

Smittytron opened the pull request.

Smittytron committed:
Add Counterstrike mission Fall of Greece 1: Personal War

Mailaender commented on mods/ra/maps/fall-of-greece-1-personal-war/personal-war.lua:
```diff
@@ -0,0 +1,492 @@
+--[[
+   Copyright 2007-2018 The OpenRA Developers (see AUTHORS)
```
2020

Mailaender commented on mods/ra/maps/fall-of-greece-1-personal-war/personal-war.lua:
```diff
@@ -0,0 +1,492 @@
+			actor.Type == "gnrl.noautotarget"
+	end)
+
+	Trigger.OnAnyKilled(VIPs, function()
```
You shouldn't do this at every tick. Do the trigger setup once after all player actors are spawned.

Mailaender commented on mods/ra/maps/fall-of-greece-1-personal-war/personal-war.lua:
```diff
@@ -0,0 +1,492 @@
+FootprintTrigger6 = { CPos.New(81, 57), CPos.New(81, 58), CPos.New(81, 59), CPos.New(81, 60), CPos.New(81, 61) }
+FootprintTrigger7 = { CPos.New(73, 58), CPos.New(73, 59), CPos.New(73, 60), CPos.New(73, 61) }
+FootprintTrigger8 = { CPos.New(51, 83), CPos.New(51, 84), CPos.New(51, 85), CPos.New(51, 86), CPos.New(51, 87), CPos.New(51, 88) }
+FootprintTrigger9 = { CPos.New(28, 77), CPos.New(29, 77), CPos.New(30, 77),  CPos.New(31, 77),  CPos.New(32, 77), CPos.New(33, 77), CPos.New(34, 77), CPos.New(35, 77), CPos.New(36, 77), CPos.New(37, 77), CPos.New(38, 77) }
```
Maybe give them some more descriptive names?

Smittytron replied:
I can if you insist. This is a very linear mission, only FootprintTrigger9 and 10 can be swapped in order. I think it is more readable to have everything numbered as almost all of the action in this mission is caused by these footprint triggers.

Mailaender replied:
I don't insist. It was more of a suggestion. If the order is important than numerical is probably the best way.

Smittytron commented:
Updated. One thing I didn't find a way to polish is the Mig that comes in and attack the barrel. For whatever reason it passes the target once before attacking it.

Mailaender commented on mods/ra/maps/fall-of-greece-1-personal-war/personal-war.lua:
```diff
@@ -0,0 +1,496 @@
+	local Mig = Reinforcements.Reinforce(ussr, { "mig" }, MigEntryPath)
+	if not BridgeBarrel1.IsDead then
+		Utils.Do(Mig, function(mig)
+			mig.Attack(BridgeBarrel1)
```
Try a forced attack:

```suggestion
			mig.Attack(BridgeBarrel1, true, true)
```

Smittytron replied:
Got the same result.

Mailaender replied:
Try setting a waypoint so it flies towards it and do the `Attack` when it sees the target.

Smittytron replied:
Got it to work by giving ussr a camera there

Smittytron commented:
Updated

Smittytron commented:
Updated by moving civilian rally to the barrels, adding a doomed extraction chopper at the outpost, and adding migs to attack them respectively.
Also varied up my civilian types in anticipation of #18099

Smittytron commented:
Rebased and updated to work with #18068 and #18126

Smittytron commented:
Rebased and Updated for #18271

Also found a place to use the new `Panic()` function

abcdefg30 commented on mods/ra/maps/fall-of-greece-1-personal-war/rules.yaml:
```diff
@@ -0,0 +1,91 @@
+
+SS:
+	AutoTarget:
+		InitialStanceAI: AttackAnything
```
This probably isn't needed any more after #18237.

abcdefg30 commented on mods/ra/maps/infiltration/rules.yaml:
```diff
@@ -68,6 +68,9 @@ SPY.Strong:
 	Passenger:
 
+DOG:
+	-IgnoresDisguise:
```
What is this change about? (I probably missed some discussion here?)

Smittytron replied:
Will remove 🤦. I was trying to test an issue with Infiltration by letting the spy speedrun through it and somehow saved the change and pushed it here. 🥇

abcdefg30 commented on mods/ra/maps/fall-of-greece-1-personal-war/personal-war.lua:
```diff
@@ -0,0 +1,521 @@
+IdleHunt = function(actor) if not actor.IsDead then Trigger.OnIdle(actor, actor.Hunt) end end
+
+MissionStart = function()
+	if not FlareBoy.IsDead then
```
Can this actor be dead at this point? If not it would be better to remove this check (we don't check for that in the trigger below either).

abcdefg30 commented on mods/ra/maps/fall-of-greece-1-personal-war/personal-war.lua:
```diff
@@ -0,0 +1,521 @@
+			end)
+
+			Trigger.AfterDelay(DateTime.Seconds(5), function()
+				FindVIPs()
```
This looks a bit odd. Why don't we use the `Trigger.OnPassengerExited` here (see https://github.com/OpenRA/OpenRA/blob/86305879cb6d8a059137598b9072dd2b03210198/mods/ra/maps/allies-03a/allies03a.lua#L76-L81 for an example) and then use `OnKilled` triggers on each passenger to fail the objective?

Smittytron replied:
I need this to work with the SendExtractionHelicopter function. Is it possible to apply a global tag with OnPassengerExited?

abcdefg30 replied:
Yes, you could add each exiting passenger to `VIPs`. Starting with `VIPs = { }` outside, and then inside the callback `VIPs[#VIPs + 1] = p`.

abcdefg30 commented on mods/ra/maps/fall-of-greece-1-personal-war/personal-war.lua:
```diff
@@ -0,0 +1,521 @@
+			end)
+
+			local guide = Actor.Create("c3", true, { Owner = greece, Location = GuideSpawn.Location })
+			if not guide.IsDead then
```
Can we remove this check? There is no way this actor can die between it being created in the line above and the line below.

abcdefg30 commented on mods/ra/maps/fall-of-greece-1-personal-war/personal-war.lua:
```diff
@@ -0,0 +1,521 @@
+			Trigger.RemoveFootprintTrigger(id)
+			foot8Triggered = true
+
+			local trig8camera = Actor.Create("camera", true, { Owner = allies, Location = TeslaCam.Location})
```
Style nit: Missing space in front of the closing brace.

abcdefg30 commented on mods/ra/maps/fall-of-greece-1-personal-war/personal-war.lua:
```diff
@@ -0,0 +1,521 @@
+			Trigger.RemoveFootprintTrigger(id)
+			foot1Triggered = true
+
+			local trig1cam = Actor.Create("camera", true, { Owner = allies, Location = Trigger1Cam.Location})
```
Same here. Actually all the camera actors seem to have that style error.

abcdefg30 commented on mods/ra/maps/fall-of-greece-1-personal-war/personal-war.lua:
```diff
@@ -0,0 +1,521 @@
+		Media.DisplayMessage(p.GetObjectiveDescription(id), "New " .. string.lower(p.GetObjectiveType(id)) .. " objective")
+	end)
+
+	ussrObj = ussr.AddPrimaryObjective("Kill Stavros.")
```
This should use the new `AddObjective` api. While we're at it, it would be good if we could have the objectives and player names start with a capital letter like all other global variables. (It wouldn't be a blocker though if you didn't want to change the player names now, but that is something to keep in mind for when porting the other missions.)

abcdefg30 reviewed (commented):
Remarks from playing and comparing with the original (remaster) version:
- Stavros ranks up instantly to the general rank, that is a bit weird. Fwiw, he also tanks a bit more in the original, so we can think about increasing his HP a bit more, but I actually like the current version, so I'm not really set here.
- Tanya can die in the original without the mission failing. I like the behaviour here (also that you have to evacuate her) more though, so that can imho stay.
- You get "Abomb launch detected" notifications when a MiG is spawned (i.e. at the map edge, not when it drops the bomb).
- The second abomb on the base is a bit delayed, so you have some more time to react (in addition to the speech notification).
- The mission actually sends 5 Medium Tanks as reinforcements. On non-easy the transport is sometimes destroyed before they are all unloaded, so you won't always see it. Considering that there are more units on normal and hard anyway (and we don't have an easy right now), 5 tanks on all difficulties sound still fair to me.
- There is a trigger that sends paratroopers to the village that is later nuked. Only after you cross another trigger the nuke is dropped on the village (with an announcement again).
- The Mammoth Tanks on the bridge leave husks in the water (we can defer that to a follow-up fix in the husk code if you don't want to work around it here), which looks odd:
![grafik](https://user-images.githubusercontent.com/7704140/85437037-42db3d00-b58a-11ea-8a84-1a6d8868d7ae.png)
- Something to think about: I think the barrels in the west are meant to destroy the escaping truck, but that doesn't work in the originals either. We might want to script it to be destroyed if it is near the barrels when they explode, or better, adjust the barrel damage and truck armor (might need a special type?) to make it blow up that way. Then we can also get rid of the exit blocking (which is how I usually get it destroyed, by placing infantry on the exit cell). If you're not comfortable with attempting this you can leave it to me.
- The Advanced Power Plant next to the barrels should blow up when the barrels do. That still doesn't put the base in low power, but the original didn't either. We can leave that to the easy difficulty imho.
- Why are there 4 flares for the helicopter?
- The helicopter should not take in Rifle Men as passengers. Otherwise you can save the men dropped for the Tesla Coil (using tanks) and put them all in there.
- The extraction helicopter disappears one cell too early.
- For the hard difficulty we should think about making infantry move towards you like they do in Allies03 (and the original).

Smittytron commented:
Updated, I'll try to address @abcdefg30 list point by point, hopefully I didn't miss anything.
  
• Changed price of Stavros to 1000
      
• It didn’t make sense to me to let Tanya die here. If you are to believe this is a part of the Allied campaign timeline, she has to serve until the end of the war.
      
• Added Abomblaunchdetected notification. Feels a bit off to me as it’s a bombing and not a launch, but I guess the og had no better ideas.
      
• Added five seconds to the second nuke
      
• Transport now drops off 5 tanks
      
• Added sov paratroops to third nuke position. Nuke timer moved back
      
• Husks in water is an issue I think needs decided separately. We’ll need to decide if we want them out of river bridges too.
      
• I think barrel damage should be addressed across the board by #17905 instead of trying to have mission specific values or adding extra barrels.
      
• Why four flares? The exported file had 4 waypoints around the LZ. I thought flares would naturally go there. If you think that is overkill I can reduce to two or one.
 
• Set passenger types to keep rifleman peasants out. 
      
• Added the GetGroundAttackers function from Allies 03, though that just has them hunting OnDamaged.

abcdefg30 commented on mods/ra/maps/fall-of-greece-1-personal-war/personal-war.lua:
```diff
@@ -0,0 +1,533 @@
+				else
+					InsertionDropHard.TargetParatroopers(InsertionPoint.CenterPosition, Angle.New(892))
+					Trigger.AfterDelay(DateTime.Seconds(6), function()
+						Media.DisplayMessage("Commander, there are several civilians in the area. \nWe'll need you to call out targets.", "Tanya")
```
Style nit: No space before `\n`.

abcdefg30 commented on mods/ra/maps/fall-of-greece-1-personal-war/personal-war.lua:
```diff
@@ -0,0 +1,533 @@
+			foot12Triggered = true
+
+			Media.PlaySoundNotification(Allies, "AlertBleep")
+			Media.DisplayMessage("Stalin will pay for what he has done today! \nI will bury him with my own hands!", "Stavros")
```
Same here.

abcdefg30 commented on mods/ra/maps/fall-of-greece-1-personal-war/rules.yaml:
```diff
@@ -0,0 +1,104 @@
+	MissionData:
+		WinVideo: toofar.vqa
+		LossVideo: bmap.vqa
+		Briefing: With the fighting fast approaching Athens, Stavros has insisted that he be allowed to survey the damage in his own hometown in Greece. \n\nYou must escort Stavros through the town and to a nearby friendly outpost. Once you reach it, evacuate Stavros to safety.\n\nIf anything should happen, get Stavros to safety any way possible. He must be kept alive.
```
And here.

abcdefg30 commented:
I was asking about the flares because that is how the original looks:
![grafik](https://user-images.githubusercontent.com/7704140/85522613-2bda3080-b606-11ea-9478-620faac623e5.png)

Smittytron commented:
Updated: Ditched the flares.

abcdefg30 commented on mods/ra/maps/fall-of-greece-1-personal-war/rules.yaml:
```diff
@@ -0,0 +1,104 @@
+	RenderSprites:
+		Image: GNRL
+	Passenger:
+		CargoType: VIP
```
That is already inherited from `GNRL`.

abcdefg30 commented on mods/ra/maps/fall-of-greece-1-personal-war/rules.yaml:
```diff
@@ -0,0 +1,104 @@
+
+E7.noautotarget:
+	Passenger:
+		CargoType: VIP
```
Same situation here.

Smittytron commented:
Updated

abcdefg30 reviewed (dismissed):
Ok.

Smittytron commented:
Updated for discord comments by Orb. Added an additional 30 second delay to chasing units after you pass the friendly outpost.

abcdefg30 commented:
Merging with @Orb370's second +1.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/d7b640ff9ea19918a97c3afa6ed16950fb03e835...18351b000caec4f94f3c1fc22b4f32115d7188ca)
